File: solutions/cpp/lasagna-master/1/lasagna_master.cpp
#include "lasagna_master.h"
#include <string>
#include <vector>

namespace lasagna_master {


int preparationTime(std::vector<std::string> layers, int avgPrepTimePerLayer){
    return layers.size() * avgPrepTimePerLayer;
}

lasagna_master::amount quantities(std::vector<std::string> layers){
    const int gramsNoodlesPerLayer = 50;
    const double litersSaucePerLayer = 0.2;
    int countSauce = 0;
    int countNoodles = 0;
    
    for(std::string layer : layers){
        if(layer == "sauce"){
            ++countSauce;
        }
        else if(layer == "noodles"){
            ++countNoodles;
        }
    }
    
    return amount{countNoodles * gramsNoodlesPerLayer, countSauce * litersSaucePerLayer};  
}

void addSecretIngredient (std::vector<std::string>& myRecipe, const std::vector<std::string>& friendList){
    for(std::string& ingredient : myRecipe){
        if(ingredient == "?"){
            ingredient = friendList.back();
        }
    }
}

std::vector<double> scaleRecipe(const std::vector<double> amount, int numOfPortions){
    std::vector<double> scaled;
    for(double quantity : amount){
        scaled.push_back((quantity/2) * numOfPortions);
    }
    return scaled;
}

void addSecretIngredient (std::vector<std::string>& myRecipe, std::string whisper){
    for(std::string& ingredient : myRecipe){
        if(ingredient == "?"){
            ingredient = whisper;
        }
    }
}
}  // namespace lasagna_master

File: solutions/javascript/lasagna-master/1/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Implement the functions needed to solve the exercise here.
 * Do not forget to export them so they are available for the
 * tests. Here an example of the syntax as reminder:
 *
 * export function yourFunction(...) {
 *   ...
 * }
 */

export function cookingStatus(remainingTime){
  if (remainingTime === 0){
    return 'Lasagna is done.';
  }
  else if (remainingTime == undefined){
    return 'You forgot to set the timer.';
  }
  else{
    return 'Not done, please wait.';
  }
}

export function preparationTime(layers, avgPrepTimePerLayer = 2){
  return layers.length * avgPrepTimePerLayer;
}

export function quantities(layers){
  let noodleQuantity = 0;
  let sauceQuantity = 0;

  for (let layer of layers) {
    if (layer == 'noodles'){
      noodleQuantity += 50;
    }
    else if (layer == 'sauce'){
      sauceQuantity += 0.2;
    }
  }

  return {'noodles': noodleQuantity, 'sauce': sauceQuantity};
}

export function addSecretIngredient(friendIngredients, myIngredients){
  myIngredients.push(friendIngredients[friendIngredients.length - 1]);
}

export function scaleRecipe(recipe, numPortions){
  const scaledRecipe = {};
  let num = numPortions/2;

  for (let key in recipe) {
    scaledRecipe[key] = recipe[key] * num;
  }

  return scaledRecipe;
}